Fix `Input::remove_joy_mapping`

MJacred opened the pull request:
Erasing a joypad mapping can invalidate other attached joypads and the fallback mapping guid

Fixes #91257 (cannot confirm crash on Linux. Untested on Windows where it was observed)

If this PR is merged, it needs to be cherry-picked/backported to all supported Godot versions, afaik.

**NOTE**
* ~~I'm not sure if I handled `fallback_mapping` correctly, because according to `Input::is_joy_known` the fallback does not count as a known device… Therefore, feedback on this would be nice~~
  * ~~If it's only meant as an internal fallback, a more minimal change of this PR to the fallback logic would be to deny removing the mapping for that one. Then I could remove the setter and getter for the fallback.~~ -> **Removed bindings**
* ~~I renamed some members in windows joypad file to differentiate between direct INPUT and direct X. I found `d_joypads`, and it feels like the `d_` prefix was meant as a direct input abbreviation. So I used that. I'm open to alternative naming convention, though. Some renaming should happen to untangle INPUT and X.~~ -> **Renaming is done in another PR**

Mickeon commented on doc/classes/Input.xml:
```diff
@@ -309,6 +315,7 @@
 			<param index="0" name="guid" type="String" />
 			<description>
 				Removes all mappings from the internal database that match the given GUID.
+				If you didn't set a GUID fallback mapping (see [method get_fallback_mapping]), then this only affects Android: If the GUID is used as a fallback mapping, the fallback will be invalidated. Use [method get_fallback_mapping] before calling this function to check your current setting. Call [method set_fallback_mapping] to set a new fallback GUID, which must be known (see [method is_joy_known]).
```
Not sure if this is absolutely necessary, but either way:
```suggestion
				[b]Note:[/b] If you didn't set a GUID fallback mapping (see [method get_fallback_mapping]), then this method only affects Android: If the GUID is used as a fallback mapping, the fallback will be invalidated. Use [method get_fallback_mapping] before calling this function to check your current setting. Call [method set_fallback_mapping] to set a new fallback GUID, which must be known (see [method is_joy_known]).
```

MJacred replied:
I shortened it. Hope that helps

MJacred commented on doc/classes/Input.xml:
```diff
@@ -140,7 +146,7 @@
 			<param index="0" name="device" type="int" />
 			<description>
-				Returns the name of the joypad at the specified device index, e.g. [code]PS4 Controller[/code]. Godot uses the [url=https://github.com/gabomdq/SDL_GameControllerDB]SDL2 game controller database[/url] to determine gamepad names.
+				Returns the name of the joypad at the specified device index, e.g. [code]PS4 Controller[/code]. Godot uses the [url=https://github.com/gabomdq/SDL_GameControllerDB]SDL2 game controller database[/url] to determine gamepad names. Although Godot overwrites [url=https://github.com/godotengine/godot/blob/master/core/input/godotcontrollerdb.txt]some of the names[/url], especially on Windows for XIn ...
```
"\_\_XINPUT_DEVICE__" is hardcoded in `joypad_windows.cpp`. That's why all xinput devices find the "GUID" called "\_\_XINPUT_DEVICE__" in godotcontrollerdb.txt

MJacred committed:
Fix Input::remove_joy_mapping

Erasing a joypad mapping can invalidate other attached joypads and the fallback mapping guid

MJacred commented:
status update:
* Rebased PR to latest master, resolving merge conflict.
* Removed bindings in Godot's spirit of "don't give bindings to things people don't need"

akien-mga commented:
CC @Rindbee if you can review this as you're also familiar with that code after #95486.

Rindbee commented on core/input/input.cpp:
```diff
@@ -1712,16 +1712,44 @@ void Input::add_joy_mapping(const String &p_mapping, bool p_update_existing) {
 }
 
 void Input::remove_joy_mapping(const String &p_guid) {
+	int fallback_mapping_offset = 0; // Fix the fallback, if we invalidate its index.
```
```suggestion
	int index_removed = 0;
```
I think it might be simpler and more intuitive to use a variable to record the index position of the removal.

The old index can be used for comparison later. `index` greater than this value needs to be `index - 1`; `fallback_mapping` needs to be reset to `-1` if equal. Double loop can be avoided.

MJacred replied:
I applied your feedback.

Please note that when adding mappings and when connecting joypads, the logic respects multiple mappings for the same GUID. It does not take the first match, but rather the last.  
With this change you requested, in the event that a guid with said multiple mappings is deleted, the index correction is not necessarily correct. The offset can be between 0 up to the amount of mappings which were deleted.

Rindbee replied:
There seems to be only one mapping for the same guid, but `fallback_mapping` can be shared by devices with other guids.

You can mark the case of removing `fallback_mapping`.  Then handle this situation in the following loop. The previous code handles this situation, so I think your previous code logic is not wrong, just not concise enough.

Rindbee replied:
Sorry, my mistake, a guid can have multiple mappings. Updating the index after the first mapping record of `p_guid` requires a second loop.

MJacred replied:
No worries :)

Rindbee reviewed (commented):
I find it easier and more intuitive to record the old index than to calculate it. There is nothing wrong with the logic.

MJacred committed:
Apply feedback + Remove unused variable `uid` in parse_mapping()

Rindbee commented on core/input/input.cpp:
```diff
@@ -1712,15 +1709,28 @@ void Input::add_joy_mapping(const String &p_mapping, bool p_update_existing) {
+		} else if (joy.mapping > index_removed) {
+			// The map_db update offset this joypad's mapping reference, update it:
+			_set_joypad_mapping(joy, map_db[joy.mapping - 1]);
 		}
```
```suggestion
		} else if (fallback_mapping_removed && joy.mapping == index_removed  ){
			_set_joypad_mapping(joy, -1);
		}
```

Rindbee replied:
```
if (fallback_mapping > index_removed ){
    fallback_mapping -= 1;
}
```

bruvzg commented on core/input/input.cpp:
```diff
@@ -1712,15 +1709,28 @@ void Input::add_joy_mapping(const String &p_mapping, bool p_update_existing) {
 			_set_joypad_mapping(joy, -1);
+		} else if (joy.mapping > index_removed) {
+			// The map_db update offset this joypad's mapping reference, update it:
+			_set_joypad_mapping(joy, map_db[joy.mapping - 1]);
```
I assume this supposed to be:
```suggestion
			_set_joypad_mapping(joy, joy.mapping - 1);
```
Since it's taking index, not mapping.

MJacred replied:
Err, that's right. Thanks for catching that!

Rindbee commented on core/input/input.cpp:
```diff
@@ -1712,15 +1709,28 @@ void Input::add_joy_mapping(const String &p_mapping, bool p_update_existing) {
+		} else if (joy.mapping > index_removed) {
+			// The map_db update offset this joypad's mapping reference, update it:
+			_set_joypad_mapping(joy, map_db[joy.mapping - 1]);
 		}
```
```suggestion
		} else if (joy.mapping == index_removed ){
			_set_joypad_mapping(joy, fallback_mapping);
		}
```

MJacred replied:
This is equal to `if (joy.puid == _guid)`, so there should be no need for that else-if-statement

MJacred committed:
Apply feedback

Rindbee commented:
@MJacred Things are more complicated than I initially thought. Especially in the case of being between the minimum removal index and the maximum removal index.  For example, suppose that more than three indexes are removed, and there is another guid between the minimum removed index and the maximum removed index, and the guid has multiple mipping records.
```
--------r0--ai-----r1-aj---r2----r3------
```

Using simple variables is not enough to record these shifts. To do this, we may need to record the various mapping indexes of the `p_guid`.

MJacred commented:
@Rindbee: I think I got it.

Alternatively, I can iterate over `joy_names` in the first for-loop, right where I remove the entry and fix the offset for each affected joypad right then and there

MJacred committed:
Updating the index after the first mapping record of p_guid requires a second loop

Rindbee commented on core/input/input.cpp:
```diff
@@ -1712,15 +1709,53 @@ void Input::add_joy_mapping(const String &p_mapping, bool p_update_existing) {
+				// Re-validate the joypad's correct mapping. Fix it if necessary.
+				int mapping = fallback_mapping;
+				for (int i = 0; i < map_db.size(); i++) {
+					if (joy.uid == map_db[i].uid) {
+						mapping = i;
+					}
```
This is actually equivalent to using the last mipping record. This may modify the previous mapping. 
Because `add_joy_mapping()` can choose not to update the existing mapping.

```
offset    -1            -2          -3              -4
----------r0----ai--aj--r1---am------r2-----an------r3--------------
```
```
--------------ai--aj-----am-----------an--------------------
```

I suggest using an array `removed_idx` to record the removed indexes (`r0`, `r1`, .. `r3`), and then compare the removed indexes here.

```cpp
	if (joy.mapping < min_removed_idx) {
		continue; // Not affected.
	}

	if (joy.mapping > max_removed_idx) {
		_set_joypad_mapping(joy, joy.mapping - removed_idx_size);
		continue; // Simple offset.
	}

	for (int i = 0; i < removed_idx_size; i++) {
		if (removed_idx[i] == joy.mapping) {
			_set_joypad_mapping(joy, fallback_mapping); // fallback_mapping need update first.
			break; // Removed, including the case where it is used as fallback_mapping.
		}
		if (removed_idx[i] < joy.mapping) {
			_set_joypad_mapping(joy, joy.mapping - (removed_idx_size - i));
			break; // Affected indexes.
		}
	}
```

MJacred committed:
Ue an array removed_idx

Rindbee reviewed (approved):
There is no problem with the logic and it is relatively clear.

Repiteo commented:
Thanks!
